### PR TITLE
WT-9199 Checkpoint can miss transactions with commit times before stable

### PR DESCRIPTION
**Pull request by raviprakashgiri29**
This pull request has the following changes.

- A csuite Test to reproduce the checkpoint can miss transactions with commit times before stable.
- Added a timing stress to the commit transaction.
- Added a new commit generation and fixed checkpoint so as not to miss transactions with commit times before stable.

### CODE REVIEW

File: src/txn/txn.c [hunk 1]
```diff
@@ -1669,13 +1669,38 @@ __txn_mod_compare(const void *a, const void *b)
...
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_TXN *txn;
+    WT_TXN_GLOBAL *txn_global;
+
+    conn = S2C(session);
+    txn = session->txn;
+    txn_global = &conn->txn_global;
+
+    if (txn_global->has_stable_timestamp && txn->first_commit_timestamp != 0 &&
```
Comment (quchenhao) [inline src/txn/txn.c:1687]: txn->first_commit_timestamp != 0 should check against WT_TS_NONE

File: src/txn/txn.c [hunk 2]
```diff
@@ -1669,13 +1669,38 @@ __txn_mod_compare(const void *a, const void *b)
...
+      txn_global->stable_timestamp >= txn->first_commit_timestamp)
+        WT_RET_MSG(session, WT_ROLLBACK,
```
Comment (quchenhao) [inline src/txn/txn.c:1689]: I think we should also update the rollback reason field.

Comment (quchenhao) [reply src/txn/txn.c:1689]: I think we probably should return EINVAL instead of WT_ROLLBACK

File: src/txn/txn.c [hunk 3]
```diff
@@ -1884,6 +1909,16 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
...
 #endif
 
+    /* Add a 2 second wait to simulate commit transaction slowness. */
+    tsp.tv_sec = 2;
+    tsp.tv_nsec = 0;
+    __wt_timing_stress(session, WT_TIMING_STRESS_COMMIT_TRANSACTION_SLOW, &tsp);
+
+    if (!prepare) {
+        WT_ERR(__txn_check_if_stable_has_moved_ahead_commit_ts(session));
+        __wt_session_gen_enter(session, WT_GEN_TXN_COMMIT);
```
Comment (quchenhao) [inline src/txn/txn.c:1919]: Why we enter the generation after the check?

File: src/txn/txn.c [hunk 4]
```diff
@@ -2009,6 +2050,7 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
 
     WT_TRET(__wt_session_reset_cursors(session, false));
     WT_TRET(__wt_txn_rollback(session, cfg));
+
```
Comment (quchenhao) [inline src/txn/txn.c:2053]: empty line

File: src/txn/txn.c [hunk 5]
```diff
@@ -1884,6 +1909,16 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
     txn->prepare_count = 0;
 #endif
 
+    /* Add a 2 second wait to simulate commit transaction slowness. */
+    tsp.tv_sec = 2;
+    tsp.tv_nsec = 0;
+    __wt_timing_stress(session, WT_TIMING_STRESS_COMMIT_TRANSACTION_SLOW, &tsp);
+
+    if (!prepare) {
+        WT_ERR(__txn_check_if_stable_has_moved_ahead_commit_ts(session));
```
Comment (quchenhao) [inline src/txn/txn.c:1918]: Can we also remove the timestamp check when we resolve each update?

File: src/txn/txn.c (hunk 1 above)
Comment (raviprakashgiri29) [reply src/txn/txn.c]: Fixed in f08e9354669258aef7f1ed7607f4960fabd9ae1b

File: src/txn/txn.c (hunk 2 above)
Comment (raviprakashgiri29) [reply src/txn/txn.c]: Fixed in f08e9354669258aef7f1ed7607f4960fabd9ae1b

File: src/txn/txn.c (hunk 3 above)
Comment (raviprakashgiri29) [reply src/txn/txn.c:1918]: @quchenhao As discussed and fixed in f08e9354669258aef7f1ed7607f4960fabd9ae1b

File: src/txn/txn.c (hunk 4 above)
Comment (raviprakashgiri29) [reply src/txn/txn.c]: Fixed in f08e9354669258aef7f1ed7607f4960fabd9ae1b

File: src/txn/txn.c [hunk 6]
```diff
@@ -1979,6 +2014,12 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
           "durable timestamp");
     }
 
+    /*
+     * Leave the commit generation after all the updates are processed and the snapshot is released.
+     */
+    if (!prepare)
+        __wt_session_gen_leave(session, WT_GEN_TXN_COMMIT);
```
Comment (quchenhao) [inline src/txn/txn.c:2021]: We can leave the generation after __wt_txn_release(session); No need to hold this longer.

Comment (quchenhao) [reply src/txn/txn.c:2021]: change the comment to snapshot is released

File: src/txn/txn.c [hunk 7]
```diff
@@ -1884,6 +1909,16 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
     txn->prepare_count = 0;
 #endif
 
+    /* Add a 2 second wait to simulate commit transaction slowness. */
+    tsp.tv_sec = 2;
+    tsp.tv_nsec = 0;
+    __wt_timing_stress(session, WT_TIMING_STRESS_COMMIT_TRANSACTION_SLOW, &tsp);
+
+    if (!prepare) {
```
Comment (quchenhao) [inline src/txn/txn.c:1917]: We need a comment here explain how this should work.

File: src/txn/txn.c (hunk 6 above)
Comment (raviprakashgiri29) [reply src/txn/txn.c]: Fixed in 9ef616b393614af333f6c1f0660018ebab2effef

File: src/txn/txn.c (hunk 7 above)
Comment (raviprakashgiri29) [reply src/txn/txn.c:1928]: Added the comment in 9ef616b393614af333f6c1f0660018ebab2effef

File: src/txn/txn.c [hunk 8]
```diff
@@ -1669,13 +1669,38 @@ __txn_mod_compare(const void *a, const void *b)
     return (0);
 }
 
+/*
+ * __txn_check_if_stable_has_moved_ahead_commit_ts --
+ *     Check if the stable timestamp has moved ahead of the commit timestamp.
+ */
+static int
+__txn_check_if_stable_has_moved_ahead_commit_ts(WT_SESSION_IMPL *session)
```
Comment (bvpvamsikrishna) [inline src/txn/txn.c:1677]: Overall change looks good to me, I think this separate function might not be required, as this is more specific to commit.

Comment (raviprakashgiri29) [reply src/txn/txn.c:1677]: @bvpvamsikrishna I prefer this to be a function for code modularization, but I am happy to change if you have a strong opinion.

**Review by quchenhao — APPROVED** [review]
lgtm

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

**Pull request merged**